sdk/web-pubsub/web-pubsub/src/hubClient.ts

```diff
@@ -231,41 +264,60 @@ export class WebPubSubServiceClient {
    */
   constructor(
     endpoint: string,
-    credential: AzureKeyCredential,
+    credential: AzureKeyCredential | TokenCredential,
     hubName: string,
-    options?: HubAdminClientOptions
+    options?: WebPubSubServiceClientOptions
   );
   constructor(
     endpointOrConnectionString: string,
-    credsOrHubName?: AzureKeyCredential | string,
-    hubNameOrOpts?: string | HubAdminClientOptions,
-    opts?: HubAdminClientOptions
+    credsOrHubName?: AzureKeyCredential | TokenCredential | string,
+    hubNameOrOpts?: string | WebPubSubServiceClientOptions,
+    opts?: WebPubSubServiceClientOptions
   ) {
     // unpack constructor arguments
-    if (typeof credsOrHubName === "object" && "key" in credsOrHubName) {
+    if (typeof credsOrHubName === "object") {
       this.endpoint = endpointOrConnectionString;
-      this.credential = credsOrHubName;
       this.hubName = hubNameOrOpts as string;
       this.clientOptions = opts;
+      if ("key" in credsOrHubName) {
+        this.credential = credsOrHubName;
+      } else {
+        this.credential = credsOrHubName;
+      }
     } else {
       const parsedCs = parseConnectionString(endpointOrConnectionString);
       this.endpoint = parsedCs.endpoint;
       this.credential = parsedCs.credential;
       this.hubName = credsOrHubName as string;
-      this.clientOptions = hubNameOrOpts as HubAdminClientOptions;
+      this.clientOptions = hubNameOrOpts as WebPubSubServiceClientOptions;
     }
 
-    const internalPipelineOptions: InternalPipelineOptions = {
+    const internalPipelineOptions: AzureWebPubSubServiceRestAPIOptionalParams = {
       ...this.clientOptions,
       ...{
         loggingOptions: {
           logger: logger.info
         }
-      }
+      },
+      ...(isTokenCredential(this.credential)
+        ? {
+            credential: isTokenCredential(this.credential) ? this.credential : undefined,
+            credentialScopes: ["https://webpubsub.azure.com/.default"]
+          }
+        : {})
```

Comment:
this is pretty fancy... but why not just put an else below and add the policy yourself instead, since you're already adding the policy for the non-token credential case?

Reply:
The credential and credentialScope is for the AAD auth policy (and so needs to be passed to the generated client constructor). The WPSKeyCredentialPolicy is for API key-based auth.

Reply:
All the generated client constructor does is have ServiceClient insert the bearerTokenAuthenticationPolicy with the scope and credential... but you can just as easily create that policy and add it too. It doesn't really matter, just seems easier to do it since you're already adding the other one